File: backend/financials.py
"""
financials.py — builds full financial statement data and computes key ratios
for the Financials tab, using pre-fetched EDGAR data.

Data source: SEC EDGAR (via EdgarFinancials).
"""

from dataclasses import dataclass
from typing import List, Optional

from .edgar_extractor import EdgarFinancials


# ── Data containers ───────────────────────────────────────────────────────────

@dataclass
class FinancialYear:
    year: int

    # ── Income Statement ──
    revenue: float
    cost_of_revenue: float
    gross_profit: float
    rd_expenses: float
    sga_expenses: float
    operating_expenses: float
    operating_income: float
    interest_expense: float
    pretax_income: float
    tax_expense: float
    net_income: float
    ebitda: float
    diluted_shares: float
    eps: float

    # ── Balance Sheet ──
    cash: float
    total_current_assets: float
    total_assets: float
    total_current_liabilities: float
    total_debt: float
    total_liabilities: float
    total_equity: float

    # ── Cash Flow ──
    operating_cash_flow: float
    capex: float
    free_cash_flow: float
    dividends_paid: float
    share_repurchases: float
    da: float

    # ── Computed Ratios ──
    # Profitability
    gross_margin: Optional[float]
    operating_margin: Optional[float]
    net_margin: Optional[float]
    roe: Optional[float]             # net income / equity
    roa: Optional[float]             # net income / total assets
    roic: Optional[float]            # NOPAT / invested capital

    # Liquidity
    current_ratio: Optional[float]   # current assets / current liabilities

    # Leverage
    debt_to_equity: Optional[float]  # total debt / equity
    interest_coverage: Optional[float]  # operating income / interest expense

    # Efficiency
    asset_turnover: Optional[float]  # revenue / total assets

    # Growth (y/y)
    revenue_growth: Optional[float]
    net_income_growth: Optional[float]
    eps_growth: Optional[float]

    # Per-share & valuation helpers
    fcf_per_share: Optional[float]


@dataclass
class FinancialsResult:
    ticker: str
    years: List[FinancialYear]


# ── Helpers ───────────────────────────────────────────────────────────────────

def _safe_div(num: float, denom: float) -> Optional[float]:
    """Return num/denom or None if denom is zero or negative."""
    if denom and denom != 0:
        return num / denom
    return None


def _safe_pct(num: float, denom: float) -> Optional[float]:
    """Return num/denom as a ratio, or None."""
    return _safe_div(num, denom)


# ── Main entry point ──────────────────────────────────────────────────────────

def fetch_financials(ticker: str, edgar_data: EdgarFinancials) -> FinancialsResult:
    """
    Build full financial statement data and compute key ratios from
    EDGAR data.

    Parameters
    ----------
    ticker     : Stock ticker symbol.
    edgar_data : Pre-fetched EDGAR financial data (oldest→newest).

    Returns
    -------
    FinancialsResult with per-year financial data and computed ratios.
    """
    ticker = ticker.upper()

    if not edgar_data.years:
        raise ValueError(f"No financial data available for {ticker}")

    years: List[FinancialYear] = []

    for ey in edgar_data.years:
        # ── Income Statement fields ──
        revenue = ey.revenue
        cost_of_revenue = ey.cost_of_revenue or 0.0
        gross_profit = ey.gross_profit or (revenue - cost_of_revenue)
        rd_expenses = ey.rd_expenses or 0.0
        sga_expenses = ey.sga_expenses or 0.0
        operating_expenses = ey.operating_expenses or 0.0
        operating_income = ey.operating_income or 0.0
        interest_expense = ey.interest_expense or 0.0
        pretax_income = ey.pretax_income or 0.0
        tax_expense = ey.income_tax or 0.0
        net_income = ey.net_income
        ebitda = ey.ebitda or (operating_income + ey.depreciation_amortization)
        diluted_shares = ey.diluted_shares
        eps = ey.eps or (net_income / diluted_shares if diluted_shares > 0 else 0.0)

        # ── Balance Sheet fields ──
        cash = ey.cash or 0.0
        total_current_assets = ey.current_assets or 0.0
        total_assets = ey.total_assets or 0.0
        total_current_liabilities = ey.current_liabilities or 0.0
        ltd = ey.long_term_debt or 0.0
        std = ey.short_term_debt or 0.0
        total_debt = ltd + std
        total_liabilities = ey.total_liabilities or 0.0
        total_equity = ey.total_equity or 0.0

        # ── Cash Flow fields ──
        operating_cash_flow = ey.operating_cash_flow or 0.0
        capex = ey.capex
        free_cash_flow = ey.fcf or (net_income + ey.depreciation_amortization - capex)
        dividends_paid = ey.dividends_paid or 0.0
        share_repurchases = ey.share_repurchases or 0.0
        da = ey.depreciation_amortization

        # ── Compute Ratios ──
        gross_margin = _safe_pct(gross_profit, revenue)
        operating_margin = _safe_pct(operating_income, revenue)
        net_margin = _safe_pct(net_income, revenue)
        roe = _safe_pct(net_income, total_equity)
        roa = _safe_pct(net_income, total_assets)

        # ROIC = NOPAT / Invested Capital
        tax_rate = tax_expense / pretax_income if pretax_income > 0 else 0.21
        nopat = operating_income * (1 - tax_rate)
        invested_capital = total_equity + total_debt - cash
        roic = _safe_pct(nopat, invested_capital) if invested_capital > 0 else None

        current_ratio = _safe_div(total_current_assets, total_current_liabilities)
        debt_to_equity = _safe_div(total_debt, total_equity) if total_equity > 0 else None
        interest_coverage = _safe_div(operating_income, interest_expense) if interest_expense > 0 else None
        asset_turnover = _safe_div(revenue, total_assets)

        fcf_per_share = _safe_div(free_cash_flow, diluted_shares)

        years.append(FinancialYear(
            year=ey.fiscal_year,
            revenue=revenue,
            cost_of_revenue=cost_of_revenue,
            gross_profit=gross_profit,
            rd_expenses=rd_expenses,
            sga_expenses=sga_expenses,
            operating_expenses=operating_expenses,
            operating_income=operating_income,
            interest_expense=interest_expense,
            pretax_income=pretax_income,
            tax_expense=tax_expense,
            net_income=net_income,
            ebitda=ebitda,
            diluted_shares=diluted_shares,
            eps=eps,
            cash=cash,
            total_current_assets=total_current_assets,
            total_assets=total_assets,
            total_current_liabilities=total_current_liabilities,
            total_debt=total_debt,
            total_liabilities=total_liabilities,
            total_equity=total_equity,
            operating_cash_flow=operating_cash_flow,
            capex=capex,
            free_cash_flow=free_cash_flow,
            dividends_paid=dividends_paid,
            share_repurchases=share_repurchases,
            da=da,
            gross_margin=round(gross_margin, 4) if gross_margin is not None else None,
            operating_margin=round(operating_margin, 4) if operating_margin is not None else None,
            net_margin=round(net_margin, 4) if net_margin is not None else None,
            roe=round(roe, 4) if roe is not None else None,
            roa=round(roa, 4) if roa is not None else None,
            roic=round(roic, 4) if roic is not None else None,
            current_ratio=round(current_ratio, 2) if current_ratio is not None else None,
            debt_to_equity=round(debt_to_equity, 2) if debt_to_equity is not None else None,
            interest_coverage=round(interest_coverage, 1) if interest_coverage is not None else None,
            asset_turnover=round(asset_turnover, 2) if asset_turnover is not None else None,
            revenue_growth=None,
            net_income_growth=None,
            eps_growth=None,
            fcf_per_share=round(fcf_per_share, 2) if fcf_per_share is not None else None,
        ))

    # Data is already oldest → newest from EDGAR

    # Compute y/y growth rates
    for i in range(1, len(years)):
        prev = years[i - 1]
        curr = years[i]
        if prev.revenue > 0:
            curr.revenue_growth = round((curr.revenue - prev.revenue) / prev.revenue, 4)
        if prev.net_income > 0:
            curr.net_income_growth = round((curr.net_income - prev.net_income) / prev.net_income, 4)
        if prev.eps > 0:
            curr.eps_growth = round((curr.eps - prev.eps) / prev.eps, 4)

    print(f"[Financials] {ticker}: {len(years)} years loaded ({years[0].year}–{years[-1].year})")
    return FinancialsResult(ticker=ticker, years=years)

File: frontend/js/financials.js
/* financials.js — Financials tab: statements, ratios, and key metrics */


/* ── Format helpers ── */

function finFmt(value, type) {
  if (value == null || isNaN(value)) return '—';
  switch (type) {
    case 'dollar':
      // Millions with 1 decimal
      if (Math.abs(value) >= 1e9) return '$' + (value / 1e9).toFixed(1) + 'B';
      if (Math.abs(value) >= 1e6) return '$' + (value / 1e6).toFixed(1) + 'M';
      if (Math.abs(value) >= 1e3) return '$' + (value / 1e3).toFixed(1) + 'K';
      return '$' + value.toFixed(0);
    case 'pct':
      return (value * 100).toFixed(1) + '%';
    case 'ratio':
      return value.toFixed(2) + 'x';
    case 'per_share':
      return '$' + value.toFixed(2);
    case 'shares':
      if (Math.abs(value) >= 1e9) return (value / 1e9).toFixed(2) + 'B';
      if (Math.abs(value) >= 1e6) return (value / 1e6).toFixed(1) + 'M';
      return value.toFixed(0);
    case 'coverage':
      return value.toFixed(1) + 'x';
    default:
      return String(value);
  }
}


/* ── Statement sub-tab state ── */

let _finActiveStatement = 'income';


/* ── Build table row ── */

function finRow(label, years, field, format, rowClass) {
  let cls = rowClass ? ` class="${rowClass}"` : '';
  let html = `<tr${cls}><td>${label}</td>`;
  for (const y of years) {
    const val = y[field];
    html += `<td>${finFmt(val, format)}</td>`;
  }
  return html + '</tr>';
}


/* ── Render: income statement table ── */

function renderIncomeStatement(years) {
  return `
    <table class="dcf-proj-table">
      <thead><tr>
        <th class="col-label">Metric</th>
        ${years.map(y => `<th>FY${y.year}</th>`).join('')}
      </tr></thead>
      <tbody>
        ${finRow('Revenue', years, 'revenue', 'dollar', 'row-bold')}
        ${finRow('Revenue Growth', years, 'revenue_growth', 'pct', 'row-italic')}
        ${finRow('Cost of Revenue', years, 'cost_of_revenue', 'dollar', 'row-cost')}
        ${finRow('Gross Profit', years, 'gross_profit', 'dollar', 'row-bold')}
        ${finRow('Gross Margin', years, 'gross_margin', 'pct', 'row-italic')}
        <tr class="section-header"><td colspan="${years.length + 1}">Operating Expenses</td></tr>
        ${finRow('R&D', years, 'rd_expenses', 'dollar', 'row-cost')}
        ${finRow('SG&A', years, 'sga_expenses', 'dollar', 'row-cost')}
        ${finRow('Total OpEx', years, 'operating_expenses', 'dollar', 'row-cost')}
        ${finRow('Operating Income', years, 'operating_income', 'dollar', 'row-bold')}
        ${finRow('Operating Margin', years, 'operating_margin', 'pct', 'row-italic')}
        <tr class="section-header"><td colspan="${years.length + 1}">Below the Line</td></tr>
        ${finRow('Interest Expense', years, 'interest_expense', 'dollar', 'row-cost')}
        ${finRow('Pretax Income', years, 'pretax_income', 'dollar', '')}
        ${finRow('Tax Expense', years, 'tax_expense', 'dollar', 'row-cost')}
        ${finRow('Net Income', years, 'net_income', 'dollar', 'row-highlight')}
        ${finRow('Net Margin', years, 'net_margin', 'pct', 'row-italic')}
        <tr class="section-header"><td colspan="${years.length + 1}">Per Share</td></tr>
        ${finRow('EPS (Diluted)', years, 'eps', 'per_share', 'row-bold')}
        ${finRow('EPS Growth', years, 'eps_growth', 'pct', 'row-italic')}
        ${finRow('Diluted Shares', years, 'diluted_shares', 'shares', '')}
        ${finRow('EBITDA', years, 'ebitda', 'dollar', 'row-bold')}
      </tbody>
    </table>
  `;
}


/* ── Render: balance sheet table ── */

function renderBalanceSheet(years) {
  return `
    <table class="dcf-proj-table">
      <thead><tr>
        <th class="col-label">Metric</th>
        ${years.map(y => `<th>FY${y.year}</th>`).join('')}
      </tr></thead>
      <tbody>
        <tr class="section-header"><td colspan="${years.length + 1}">Assets</td></tr>
        ${finRow('Cash & Equivalents', years, 'cash', 'dollar', '')}
        ${finRow('Total Current Assets', years, 'total_current_assets', 'dollar', '')}
        ${finRow('Total Assets', years, 'total_assets', 'dollar', 'row-bold')}
        <tr class="section-header"><td colspan="${years.length + 1}">Liabilities</td></tr>
        ${finRow('Total Current Liabilities', years, 'total_current_liabilities', 'dollar', '')}
        ${finRow('Total Debt', years, 'total_debt', 'dollar', '')}
        ${finRow('Total Liabilities', years, 'total_liabilities', 'dollar', 'row-bold')}
        <tr class="section-header"><td colspan="${years.length + 1}">Equity</td></tr>
        ${finRow("Shareholders' Equity", years, 'total_equity', 'dollar', 'row-bold')}
      </tbody>
    </table>
  `;
}


/* ── Render: cash flow table ── */

function renderCashFlow(years) {
  return `
    <table class="dcf-proj-table">
      <thead><tr>
        <th class="col-label">Metric</th>
        ${years.map(y => `<th>FY${y.year}</th>`).join('')}
      </tr></thead>
      <tbody>
        ${finRow('Operating Cash Flow', years, 'operating_cash_flow', 'dollar', 'row-bold')}
        ${finRow('D&A', years, 'da', 'dollar', '')}
        ${finRow('Capital Expenditures', years, 'capex', 'dollar', 'row-cost')}
        ${finRow('Free Cash Flow', years, 'free_cash_flow', 'dollar', 'row-fcf')}
        ${finRow('FCF / Share', years, 'fcf_per_share', 'per_share', 'row-italic')}
        <tr class="section-header"><td colspan="${years.length + 1}">Capital Return</td></tr>
        ${finRow('Dividends Paid', years, 'dividends_paid', 'dollar', '')}
        ${finRow('Share Repurchases', years, 'share_repurchases', 'dollar', '')}
      </tbody>
    </table>
  `;
}


/* ── Render: ratios dashboard ── */

function renderRatiosDashboard(years) {
  return `
    <table class="dcf-proj-table">
      <thead><tr>
        <th class="col-label">Ratio</th>
        ${years.map(y => `<th>FY${y.year}</th>`).join('')}
      </tr></thead>
      <tbody>
        <tr class="section-header"><td colspan="${years.length + 1}">Profitability</td></tr>
        ${finRow('Gross Margin', years, 'gross_margin', 'pct', '')}
        ${finRow('Operating Margin', years, 'operating_margin', 'pct', '')}
        ${finRow('Net Margin', years, 'net_margin', 'pct', '')}
        ${finRow('ROE', years, 'roe', 'pct', '')}
        ${finRow('ROA', years, 'roa', 'pct', '')}
        ${finRow('ROIC', years, 'roic', 'pct', '')}
        <tr class="section-header"><td colspan="${years.length + 1}">Liquidity</td></tr>
        ${finRow('Current Ratio', years, 'current_ratio', 'ratio', '')}
        <tr class="section-header"><td colspan="${years.length + 1}">Leverage</td></tr>
        ${finRow('Debt / Equity', years, 'debt_to_equity', 'ratio', '')}
        ${finRow('Interest Coverage', years, 'interest_coverage', 'coverage', '')}
        <tr class="section-header"><td colspan="${years.length + 1}">Efficiency</td></tr>
        ${finRow('Asset Turnover', years, 'asset_turnover', 'ratio', '')}
        <tr class="section-header"><td colspan="${years.length + 1}">Growth</td></tr>
        ${finRow('Revenue Growth', years, 'revenue_growth', 'pct', '')}
        ${finRow('Net Income Growth', years, 'net_income_growth', 'pct', '')}
        ${finRow('EPS Growth', years, 'eps_growth', 'pct', '')}
      </tbody>
    </table>
  `;
}


/* ── Statement sub-tab switching ── */

function switchFinStatement(which, years) {
  _finActiveStatement = which;

  // Update sub-tab buttons
  document.querySelectorAll('#fin-sub-tabs .fin-sub-btn').forEach(btn => {
    btn.classList.toggle('active', btn.dataset.stmt === which);
  });

  // Render the selected statement
  const container = document.getElementById('fin-statement-body');
  let html = '';
  switch (which) {
    case 'income':   html = renderIncomeStatement(years); break;
    case 'balance':  html = renderBalanceSheet(years);    break;
    case 'cashflow': html = renderCashFlow(years);        break;
    case 'ratios':   html = renderRatiosDashboard(years); break;
  }
  container.innerHTML = `<div class="proj-table-scroll">${html}</div>`;
}


/* ── Main entry point ── */

function renderFinancials(data) {
  const unavailable = document.getElementById('fin-unavailable');
  const content = document.getElementById('fin-content');

  if (!data.financials_available || !data.financials || data.financials.length === 0) {
    show(unavailable);
    hide(content);
    return;
  }

  hide(unavailable);
  show(content);

  const years = data.financials;

  // Wire sub-tab buttons
  document.querySelectorAll('#fin-sub-tabs .fin-sub-btn').forEach(btn => {
    btn.onclick = () => switchFinStatement(btn.dataset.stmt, years);
  });

  // Render default (income statement)
  _finActiveStatement = 'income';
  document.querySelectorAll('#fin-sub-tabs .fin-sub-btn').forEach(btn => {
    btn.classList.toggle('active', btn.dataset.stmt === 'income');
  });
  switchFinStatement('income', years);
}
